diff --git src/core/aws-net-websocket-registry.adb src/core/aws-net-websocket-registry.adb
--- src/core/aws-net-websocket-registry.adb
+++ src/core/aws-net-websocket-registry.adb
@@ -31,6 +31,7 @@ pragma Ada_2012;
 
 with Ada.Containers.Indefinite_Ordered_Maps;
 with Ada.Containers.Ordered_Maps;
+with Ada.Strings.Fixed;
 with Ada.Unchecked_Deallocation;
 
 with AWS.Config;
@@ -829,13 +830,32 @@ package body AWS.Net.WebSocket.Registry is
    -----------------
 
    function Constructor (URI : String) return Registry.Factory is
-      Position : constant Constructors.Cursor := Factories.Find (URI);
+      Position : Constructors.Cursor := Factories.Find (URI);
+      Last     : Integer := URI'Last;
+
    begin
+      --  Exact match.
+
       if Constructors.Has_Element (Position) then
          return Constructors.Element (Position);
-      else
-         return Create'Access;
       end if;
+
+      --  Pattern lookup.
+
+      loop
+         Position := Factories.Find (URI (URI'First .. Last) & "/*");
+
+         if Constructors.Has_Element (Position) then
+            return Constructors.Element (Position);
+         end if;
+
+         exit when Last < URI'First;
+
+         Last :=
+           Ada.Strings.Fixed.Index (URI, "/", Last, Ada.Strings.Backward) - 1;
+      end loop;
+
+      return Create'Access;
    end Constructor;
 
    ------------
